tensorflow/compiler/xla/service/slice_sinker.cc

```diff
@@ -0,0 +1,276 @@
+}
+
+// Generates the new user with the whole tensor and slices its output instead
+// of the true users with sliced tensor.
+// Records the stale true users and their operand-slices to prepare removing.
+void SinkSlices(const std::vector<HloInstruction*>& operands,
+    const std::vector<HloInstruction*>& users) {
+  // Generates new user.
+  const Shape shape = operands[0]->shape();
+  PrimitiveType element_type = users[0]->shape().element_type();
+  Shape new_shape = ShapeUtil::ChangeElementType(shape, element_type);
+
+  HloComputation* computation = users[0]->parent();
+  auto new_user = computation->AddInstruction(
+      users[0]->CloneWithNewOperands(new_shape, operands));
+  VLOG(10) << "Add NewUser: " << new_user->ToString();
+
+  // Replaces the true users with new user and its user-slices.
+  for (HloInstruction* user : users) {
+    const HloInstruction* operand_slice = user->operand(0);
+    // Generates user slices of new user.
+    auto user_slice = computation->AddInstruction(
+        operand_slice->CloneWithNewOperands(user->shape(), {new_user}));
+    VLOG(10) << "Add NewSlice: " << user_slice->ToString()
+             << " Replace: " << user->ToString();
+    // Replaces true users with user slices.
+    user->ReplaceAllUsesWith(user_slice);
+  }
+}
+
+}  // namespace
+
+bool SliceSinkerImpl::IsVisited(const HloInstruction* instruction) const {
+  return visited_.contains(instruction);
+}
+
+absl::optional<std::vector<HloInstruction*>>
+ SliceSinkerImpl::FindSourceOperandsOfSlicesForElementwiseOperation(
+    const HloInstruction* inst) {
+  std::vector<HloInstruction*> operands;
+  // Check operand:
+  // The inst's operand should be a slice of the true operand.
+  // The operands-vector keeps all true operands of inst.
+  for (HloInstruction* operand_slice : inst->operands()) {
+    if (operand_slice->opcode() != HloOpcode::kSlice) {
+      // If the operand is not operand-slice, returns empty vector.
+      visited_.insert(inst);
+      return absl::nullopt;
+    }
+    // The operand-slice's operand is a true operand.
+    HloInstruction* operand = operand_slice->mutable_operand(0);
+    operands.push_back(operand);
+  }
+
+  // No true operands found. returns empty vector.
+  if (operands.empty()) {
+    visited_.insert(inst);
+    return absl::nullopt;
+  }
+
+  // Check operands:
+  // True operands should have the same shape because inst is elementwise.
+  const Shape shape = operands[0]->shape();
+  for (const HloInstruction* operand : operands) {
+    // Only support element-wise now
+    if (!ShapeUtil::CompatibleIgnoringElementType(operand->shape(), shape)) {
+      visited_.insert(inst);
+      return absl::nullopt;
+    }
+  }
+  // Inst's operand-slices should be sliced from the same location of true
+  // operands.
+  const HloInstruction* operand0 = inst->operand(0);
+  for (const HloInstruction* operand : inst->operands()) {
+    if (operand0->slice_starts() != operand->slice_starts() ||
```

Comment:
This comparison is used in at least two place, can we outline this code to a routine like this:
bool SameSliceConfiguration(const HloInstruction* slice_1, const HloInstruction* slice_2)